feat: 강좌 시청 페이지 Learn 데이터 연동 구조 추가

whozbebop opened the pull request:
## 작업 개요
강좌 시청 페이지에서 사용할 Learn 전용 뷰 모델과 서비스 레이어를 추가하고,
목 데이터 기반으로 UI까지 연동 가능한 구조를 만들었습니다.

## 주요 변경 사항

### 1. Learn 전용 FE 뷰 모델 정의 (`domains/course/types/learn.ts`)
- 백엔드 `/api/courses/{courseId}` 응답 구조를 참고해 Learn 전용 타입 정의
  - `LearnLectureResource`: 강의 리소스(VIDEO/PDF) 정보
  - `LearnLecture`: 강의 단위(lectureId, duration, isPreview, orderIndex, resource 등)
  - `LearnSection`: 섹션 단위(sectionId, title, order, lectures[])
  - `LearnCourse`: 시청 페이지에서 사용할 강좌 정보 + 섹션/강의 트리 구조
- 시청 페이지에서 필요한 데이터를 한 번에 모은 뷰 모델 추가
  - `LearnPageData = { course: LearnCourse; enrollment: Enrollment | null; progress: Progress | null }`

### 2. Enrollment status 유니온 타입 정의
- 수강 상태를 문자열 상수로 통일해서 사용할 수 있도록 유니온 타입 정의
  - `ENROLLED | COMPLETED | CANCELED | EXPIRED`
- 향후 마이페이지/시청 페이지 등에서 동일한 status 타입을 재사용할 수 있도록 분리

### 3. Learn 서비스 레이어 추가 (`domains/course/services/learnService.ts`)
- `getLearnPageData(courseId: string): Promise<LearnPageData>` 추가
  - (현재는 `NEXT_PUBLIC_USE_MOCK=true` 시 `MOCK_LEARN_PAGE_DATA`를 그대로 반환)
  - 실제 API 연동 시에는 아래 3개의 API를 조합하는 형태로 확장 예정
    - `/api/courses/{courseId}` → `LearnCourse`
    - `/api/enrollments?...` → `LearnPageData['enrollment']`
    - `/api/progresses/{enrollmentId}` → `LearnPageData['progress']`
- API 응답 타입을 서비스 내부에서 간단한 Response 타입으로 감싸서 사용하도록 구조화

### 4. Learn 목 데이터 추가 (`domains/course/mocks/learn.mock.ts`)
- `LearnPageData` 구조에 맞는 목 데이터 추가
  - 강좌 정보 + 섹션/강의 + enrollment + progress 더미 데이터를 한 번에 포함
- `NEXT_PUBLIC_USE_MOCK` 환경변수로 실제 연동 전까지는 목 데이터 기반으로 시청 페이지 UI를 확인할 수 있도록 처리

### 5. 강좌 시청 페이지 클라이언트 연동 (`domains/course/pages/CourseLearnClientPage.tsx`)
- `CourseLearnClient` 컴포넌트가 `learnData: LearnPageData`를 props로 받아 렌더링하도록 수정
- `LearnPageData.course`의 섹션/강의 데이터를 바탕으로
  - 왼쪽: 비디오/PDF 플레이어 영역
  - 오른쪽: 커리큘럼(섹션/강의 리스트) 사이드바
  를 구성하도록 반영
- 기존 더미(UICourse/UILecture 기반 하드코딩 데이터)는 제거하고,
  Learn 전용 뷰 모델에서 계산한 데이터를 사용하는 구조로 변경

### 6. 서버 컴포넌트 연동 (`app/courses/[id]/learn/page.tsx`)
- 서버 컴포넌트에서 `getLearnPageData(id)` 호출 후
  결과를 `CourseLearnClient`에 넘기도록 연동
- `params.id` 기준으로 개별 강좌 시청 페이지가 렌더링되도록 라우팅 정리

## 테스트
- [x] `NEXT_PUBLIC_USE_MOCK=true` 설정 후 `/courses/{id}/learn`에서 시청 페이지 진입
- [x] 섹션 아코디언 토글 동작 확인
- [x] 강의 선택 시 비디오/ PDF 플레이어 영역 변경 확인
- [x] 브라우저 콘솔 및 서버 로그 에러 여부 확인

## 추가 확인 필요 사항
- 실제 API 연동 시 아래 부분에서 백엔드 스펙 확정 후 타입/로직 조정 예정
  - `/api/courses/{courseId}` 응답 필드명 및 섹션/강의/리소스 구조
  - `/api/enrollments` 응답에서 `courseId`, `enrollmentId`, 진도율(progressRate) 필드 매핑
  - `/api/progresses/{enrollmentId}` 응답 구조와 `Progress` 타입 매핑
- 마이페이지 쪽 enrollment 목록 목 데이터와 `LearnPageData`의 `courseId/enrollmentId`를 맞춰서
  “마이페이지 → 수강 중 강의 클릭 → 시청 페이지” 흐름을 이어서 테스트할 계획입니다.


close #66

whozbebop committed:
chore: 주석 삭제

whozbebop committed:
feat: 강좌 시청 페이지 Learn 데이터 연동 구조 추가

- LearnLectureResource/LearnLecture/LearnSection/LearnCourse 타입 정의 및 LearnPageData 구성
- MOCK_LEARN_PAGE_DATA 추가해서 강좌/수강(enrollment)/진도(progress) 더미 데이터 연동
- getLearnPageData 서비스에서 /api/courses, /api/enrollments, /api/progresses 응답 조합
- EnrollmentStatus 유니온 타입으로 정리해 수강 상태 상수화
- learn.ts에 강좌 시청 페이지용 FE 전용 뷰 모델 구조 정리

kimsuyeon0916 commented on src/domains/course/pages/CourseLearnClientPage.tsx:
```diff
@@ -1,15 +1,24 @@
-"use client";
+// src/domains/course/pages/CourseLearnClientPage.tsx
```
```suggestion
```

kimsuyeon0916 commented on src/domains/course/pages/CourseLearnClientPage.tsx:
```diff
@@ -1,15 +1,24 @@
   id: string;
   title: string;
-  type: "video" | "pdf";
+  type: 'video' | 'pdf';
```
확장자 별도의 type으로 적용하셔야 할 듯 합니다!

kimsuyeon0916 commented on src/domains/course/types/course.ts:
```diff
@@ -18,7 +18,7 @@ export type Course = {
 
   duration: number;
-  status: 'draft' | 'published' | 'hidden';
+  status: 'DRAFT' | 'PUBLISHED' | 'HIDDEN';
```
```suggestion
  status: 'DRAFT' | 'PUBLISHED' | 'DELETED';
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,52 @@
+
+export type LearnLectureResource = {
+  resourceId: number;
+  resourceType: 'VIDEO' | 'PDF';
```
```suggestion
  resourceType: ResourceType;
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,52 @@
+  lectures: LearnLecture[];
+};
+
+export type LearnCourse = {
```
```suggestion
export type CourseLearnResponse = {
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,52 @@
+  sections: LearnSection[];
+};
+
+export type LearnPageData = {
```
이름을 바꿔보는거 어떠세요?!

kimsuyeon0916 commented on src/domains/course/types/progress.ts:
enrollment 도메인은 제거하고 course 도메인 내 타입에 넣어주세요

kimsuyeon0916 replied:
이것도 learn 에다가 넣어주세요!

kimsuyeon0916 commented on src/domains/enrollment/types/enrollment.ts:
enrollment 도메인은 제거하고 course 도메인 내 타입에 넣어주세요

kimsuyeon0916 commented on src/domains/course/types/enrolledCourse.ts:
enrollment 도메인은 제거하고 course 도메인 내 타입에 넣어주세요 (이건 겹치는게 합치면서 존재할거니까 통합도 부탁드려요!)

kimsuyeon0916 replied:
이 파일 제거하시고, enrollment 내에 넣어주세요!

kimsuyeon0916 commented on src/domains/course/mocks/learn.mock.ts:
```diff
@@ -0,0 +1,141 @@
+// src/domains/course/mocks/learn.mock.ts
+import type { LearnPageData } from '@/domains/course/types/learn';
+
+export const MOCK_LEARN_PAGE_DATA: LearnPageData = {
```
데이터 타입 변경

kimsuyeon0916 commented on src/app/courses/[id]/learn/page.tsx:
```diff
@@ -1,7 +1,13 @@
+
+export default async function CourseLearnPage({ params }: LearnPageProps) {
+  const { id } = params;
+  const learnData = await getLearnPageData(id);
```
굳이 props 안쓰고 <CourseLearnClient /> 로 그대로 하고, 

  const { id } = params;
  const learnData = await getLearnPageData(id);
  
  를 CourseLearnClient 내부에 포함하셔도 될 듯 하네요!

kimsuyeon0916 commented on src/domains/course/pages/CourseLearnClientPage.tsx:
```diff
@@ -32,90 +41,78 @@ type UICourse = {
-  const course = detail.course;
-  const sections = detail.sections.map((sec) => ({
-    id: sec.id,
+function buildUICourseFromLearnData(learnData: LearnPageData): UICourse {
```
훅으로 별도로 빼서 관리하는거 추천드릴게요!

kimsuyeon0916 replied:
formatDurationSeconds는 buildUICourseFromLearnData 리팩토링(분리 및 네이밍 변경)이후에 고민해보세용!

kimsuyeon0916 commented on src/domains/course/services/learnService.ts:
base url 관련한 환경변수를 통일해야 할 듯 합니다! 또한 이런 정보는 notion/slack 공유 부탁드려요!

제가 fetchApi(권한 필요한 요청) 반영되고 나서 수정이 필수적일 듯 합니다😅

kimsuyeon0916 replied:
response/request 통신 타입을 여기다가 말고 /types에 넣어주세요!!

kimsuyeon0916 replied:
Create → createCourse
Read → getCourseDetail
Update → updateCourse
Delete → deleteCourse

기준으로 네이밍 변경 부탁드릴게용!

whozbebop committed:
Update src/domains/course/types/course.ts

Co-authored-by: Suyeon Kim <kimsuyeon_0916@naver.com>

whozbebop committed:
Update src/domains/course/pages/CourseLearnClientPage.tsx

Co-authored-by: Suyeon Kim <kimsuyeon_0916@naver.com>

whozbebop committed:
Update src/domains/course/types/learn.ts

Co-authored-by: Suyeon Kim <kimsuyeon_0916@naver.com>

whozbebop committed:
refactor: 강좌 시청 페이지 타입 및 서비스 구조 개편

- enrollment / enrolledCourse / progress 타입을 course 도메인으로 통합
- learn.ts 타입 정의 추가 및 기존 타입 정리
- CourseLearnClientPage 및 useCourseLearn 훅 로직 개선
- learnService에 API 목업 기반 데이터 페칭 로직 정리
- 중복 타입 파일(enrollment 도메인) 제거 및 구조 일원화

whozbebop committed:
Merge branch 'feature/#66' of https://github.com/20250915-POTENUP-FE-1/LXP_lern-p into feature/#66

whozbebop committed:
강좌 시청 페이지 스타일 수정

whozbebop committed:
chore: 강좌 시청 페이지 스타일 수정

whozbebop committed:
Merge branch 'feature/#66' of https://github.com/20250915-POTENUP-FE-1/LXP_lern-p into feature/#66

kimsuyeon0916 commented on src/domains/course/types/enrollment.ts:
해당 타입들을 다 learn에다가 넣으세요!

kimsuyeon0916 commented on src/domains/course/services/learnService.ts:
```diff
@@ -0,0 +1,82 @@
+  ProgressResponse,
+} from '@/domains/course/types/learn';
+
+const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? '';
```
```suggestion
const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL ?? '';
```

kimsuyeon0916 commented on src/domains/course/utils/createLearnPageUIModel.ts:
useCourseLearn 내부 하단에 넣으세용!

kimsuyeon0916 replied:
+ 네이밍 직관적 변경

kimsuyeon0916 commented on src/domains/course/hooks/useCourseLearn.ts:
```diff
@@ -0,0 +1,70 @@
+};
+
+export function useCourseLearn(learnData: LearnPageData): UseCourseLearnReturn {
+  const courseData = useMemo(() => createLearnPageUIModel(learnData), [learnData]);
```
createLearnPageUIModel 을 분리하지 말고 useMemo 내에 넣으세용!

whozbebop committed:
refactor: Learn 도메인 타입 통합 및 Enrollment 모델 구조 정리

### 주요 변경사항
- learn.ts 도메인 모델 정리 (Domain / Response DTO / UI Model 명확화)
- Enrollment 타입을 수강 도메인에 맞게 재정의
  - `id` → `enrollmentId`로 변경
  - `createdAt` 제거, `enrolledAt` / `updatedAt` 필드 추가
  - `progress` 필드 추가하여 학습 진척도 정보 일원화
- 기존 `EnrolledCourse`, `EnrollmentCourse`, `progress` 등 중복/혼란 타입 제거
- LearnPage 라우트(params) Next.js 16 규칙에 맞게 `await params` 방식으로 수정
- learnService, useCourseLearn 훅, ClientPage 로직 타입 대응 수정
- course mock 데이터 구조를 새로운 Enrollment 타입에 맞게 업데이트

kimsuyeon0916 commented on src/app/courses/[id]/learn/page.tsx:
```diff
@@ -1,7 +1,13 @@
-  );
+type LearnPageProps = {
+  params: { id: string };
+};
```
```suggestion
type CourseLearnPageProps = {
  params: Promise<{ id: string }>;
};
```

kimsuyeon0916 commented on src/app/courses/[id]/learn/page.tsx:
```diff
@@ -1,7 +1,13 @@
+
+export default async function CourseLearnPage({ params }: LearnPageProps) {
+  const { id } = await params;
+  const learnData = await getLearnPageData(id);
```
다시 CourseLearnClientPage 내부로 넣어주세용(api 제대로 연동하고 수정 가능성 있긴해용!)

kimsuyeon0916 commented on src/domains/course/pages/CourseLearnClientPage.tsx:
```diff
@@ -1,134 +1,33 @@
+    handleLectureClick,
+    totalLectures,
+    completedLectures,
+  } = useCourseLearn(learnData);
```
learnData를 불러오는 서비스 함수를 useCourseLearn 내부에서 호출하세요!

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,149 @@
+// 구조 규칙
+// 1) Domain Model: 백엔드 비즈니스 엔티티 그대로 표현
+// 2) Response DTO: HTTP 응답 구조(래퍼 포함)
+// 3) UI Model: 프론트에서 화면 렌더링용 가공 모델
```
```suggestion

```

kimsuyeon0916 commented on src/domains/course/utils/createLearnPageUIModel.ts:
```diff
@@ -0,0 +1,47 @@
+  };
+}
+
+export function formatDurationSeconds(seconds: number) {
```
```suggestion
function formatDurationSeconds(seconds: number) {
```

kimsuyeon0916 commented on src/domains/course/hooks/useCourseLearn.ts:
```diff
@@ -0,0 +1,87 @@
+  completedLectures: number;
+};
+
+export function useCourseLearn(learnData: CourseLearn): UseCourseLearnReturn {
```
```suggestion
import type { CourseLearn } from '@/domains/course/types/learn';

export function useCourseLearn() {
```

kimsuyeon0916 replied:
이렇게 인자가 없도록 수정하시고, 아래의 로직들 타입 해당 파일 내에 정의 부탁드려요!

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,149 @@
+  instructor: string;
+  description: string;
+  sections: UISection[];
+};
```
차라리, useCourseLerarn 내에 옮기시는게 좋아보입니다(useCourseLerarn 내에 댓글 리뷰 참고하면 이 타입들이 사라지실 것 같습니다)

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
api 명세 내의 data 에 해당하는 부분만 response/request로 하기

kimsuyeon0916 commented on src/domains/course/utils/createLearnPageUIModel.ts:
파일 제거 부탁 드려요~

kimsuyeon0916 commented on src/domains/user/types/enrollment.ts:
히스토리용) 마이페이지 개발 시, 수정될 가능성 매우 높음 => 그때 확인

kimsuyeon0916 commented on src/domains/course/services/learnService.ts:
Response 타입 변경(pi 명세 내의 data 에 해당하는 부분만 response/request로 하기) 이후에 수정 무조건 필요할 듯 합니다

whozbebop committed:
feat: 강좌 시청 도메인 및 응답 타입 정리
- Learn 도메인 타입 정의
  - LearnLecture / LearnSection / LearnCourse / LearnProgress / CourseLearn 추가
- ApiResponse<T> 도입
  - 기존 status/code/message 제거
  - CourseDetailResponse / EnrollmentListResponse / ProgressResponse를 ApiResponse<T> 기반으로 정리
- Enrollment 도메인 타입 정리
  - Enrollment / EnrollmentWithCourse 구조 명확화

whozbebop committed:
feat: 강좌 시청 페이지 훅 및 클라이언트 구조 리팩토링
- learnService에서 Promise.all로 강좌/수강/진행도 병렬 호출하도록 변경
- useCourseLearn 훅에서
  - useParams로 courseId 조회
  - getLearnPageData 호출 및 상태 관리
  - 도메인 모델 → 화면용 데이터 매핑 로직 분리
- formatLecture 유틸 추가
  - mapLecture / mapSection / mapCourse로 도메인 → UI 변환 공통화
  - ReturnType을 활용해 별도 UI 타입명 없이 타입 추론
- CourseLearnClient 컴포넌트에서
  - props 없이 훅만 사용하도록 변경
  - 로딩/초기 currentLecture null 처리
  - 기존 CourseLearnPage.module.css 클래스 네이밍 유지

whozbebop committed:
Merge branch 'develop' of https://github.com/20250915-POTENUP-FE-1/LXP_lern-p into feature/#66

whozbebop committed:
feat: 강의 시청 페이지 타입/서비스/훅 구조 리팩토링

- learn.ts 타입 스키마 재정비(LearnCourse/LearnSection/LearnLecture/Progress 통합 정리)
- enrollment.ts 타입 구조 재정비 및 Learn 페이지 연동을 위한 최소 필드만 노출하도록 수정
- learnService.ts 내 API 응답 구조에 맞게 Response 타입 및 fetch 로직 일원화
- useCourseLearn 훅에서 getLearnPageData 제거 → 서비스 조합 로직 자체적으로 구성하도록 변경
- mapCourse/mapLecture 기반의 UI-friendly 데이터 변환 구조 도입

whozbebop committed:
refactor: learn.ts 타입 구조 백엔드 스펙과 일치하도록 리팩토링

kimsuyeon0916 commented on src/domains/course/pages/CourseLearnClientPage.tsx:
```diff
@@ -144,78 +43,122 @@ export default function CourseLearnPage() {
+              )}
               <span className={styles['course-learn__progress']}>
-                {currentLecture.type === "video" ? "영상 강의" : "PDF 자료"}
+                {currentLecture.type === 'VIDEO' ? '영상 강의' : 'PDF 자료'}
```
```suggestion
                {currentLecture.type === 'VIDEO' ? '영상 강의' : '강의 자료'}
```

kimsuyeon0916 commented on src/domains/course/services/learnService.ts:
```diff
@@ -0,0 +1,30 @@
+    cache: 'no-store',
+  });
+
+  return res ?? null;
```
```suggestion
  return await getApi<LearnEnrollmentResponse>(`${BASE_URL}/api/enrollments/${enrollmentId}`, {
    cache: 'no-store',
  });
```

kimsuyeon0916 commented on src/domains/course/services/learnService.ts:
```diff
@@ -0,0 +1,30 @@
+    cache: 'no-store',
+  });
+
+  return res ?? null;
```
```suggestion
  return await getApi<LearnProgressResponse>(`${BASE_URL}/api/progresses/${enrollmentId}`, {
    cache: 'no-store',
  });
```

kimsuyeon0916 commented on src/domains/course/utils/mapCourse.ts:
mapCourse.ts로 파일명 바꾸고, 함수명도 바꿔주세요!

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,85 @@
+
+/* ---------------------------------------------
+ * 4. Combined Learn Page Model
+ * --------------------------------------------- */
```
```suggestion
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,85 @@
+
+/* ---------------------------------------------
+ * 3. Progress (단건 조회)
+ * --------------------------------------------- */
```
```suggestion
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,85 @@
+
+/* ---------------------------------------------
+ * 2. Enrollment (단건 조회)
+ * --------------------------------------------- */
```
```suggestion
```

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,85 @@
+
+/* ---------------------------------------------
+ * 1. Lecture / Section / Course Domain Models
+ * --------------------------------------------- */
```
```suggestion
```

whozbebop committed:
주석 삭제

Co-authored-by: Suyeon Kim <kimsuyeon_0916@naver.com>

whozbebop committed:
refactor: LearnService API 경로 정리 및 BASE_URL 제거
- NEXT_PUBLIC_BASE_URL 사용 제거

- getCourse / getEnrollment / getProgress API 경로를 /api/... 형태로 통일

- getApi 호출 방식 단순화 및 불필요한 분기 제거

- 서버·클라이언트 환경에서의 fetch 오류 예방을 위한 구조 개선

whozbebop committed:
Merge branch 'feature/#66' of https://github.com/20250915-POTENUP-FE-1/LXP_lern-p into feature/#66

whozbebop committed:
refactor: learn.ts 아이디(id) 스트링 타입으로 변경,
enrollment: 도메인 타입 정리

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,74 @@
+
+  isPurchased: boolean;
+  studentCount: number;
+  totalduration: number;
```
동훈님 왈) `totalDuration` 라고 합니다

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
```diff
@@ -0,0 +1,74 @@
+export type LearnLecture = {
+  lectureId: string;
+  title: string;
+  duration: number;
```
👀 (노션에는 `totalDurationSeconds` 로 되어있음)

kimsuyeon0916 commented on src/domains/course/types/learn.ts:
👀 히스토리용) 강좌 상세 조회 응답 값 타입명을 course.ts 에 정의하는 것으로 추후에 수정. -> mapCourse 내 파라미터 타입도 course.ts 참조

whozbebop committed:
refactor: mock 테스트 관련 소스 삭제, id 타입 string으로 변경

whozbebop committed:
refactor: learn type 변경

kimsuyeon0916 commented on src/domains/course/hooks/useCourseLearn.ts:
```diff
@@ -0,0 +1,89 @@
+
+      // 수강 정보 단건 조회
+      const enrollment: LearnEnrollmentResponse | null =
+        course.isPurchased && course.studentCount > 0 ? await getEnrollment(courseId) : null;
```
👀 히스토리용) getEnrollment의 인자로 courseId말고 enrollmentid를 받아야 하는데, 여기 내부 로직이 조금 잘못되었습니다.

근데 막상 해당하는 enrollmentId를 가져오기 위해서는 내 수강목록 조회 API를 호출하고 이에 해당하는 courseId를 찾아서 enrollmentId를 얻어와야 할 듯 합니다.
